app/backend/src/couchers/servicers/references.py

```diff
@@ -55,11 +57,23 @@ def ListReferences(self, request, context):
             if not request.from_user_id and not request.to_user_id:
                 context.abort(grpc.StatusCode.INVALID_ARGUMENT, errors.NEED_TO_SPECIFY_AT_LEAST_ONE_USER)
 
+            users1 = aliased(User)
+            users2 = aliased(User)
             query = session.query(Reference)
             if request.from_user_id:
-                query = query.filter(Reference.from_user_id == request.from_user_id)
+                # join the to_users, because only interested if the recipient is visible
+                query = (
+                    query.join(users1, users1.id == Reference.to_user_id)
+                    .filter(~users1.is_banned)  # instead of is_visible; if user is deleted, reference still visible
+                    .filter(Reference.from_user_id == request.from_user_id)
+                )
             if request.to_user_id:
-                query = query.filter(Reference.to_user_id == request.to_user_id)
+                # join the from_users, because only interested if the writer is visible
+                query = (
+                    query.join(users2, users2.id == Reference.from_user_id)
+                    .filter(~users2.is_banned)  # instead of is_visible; if user is deleted, reference still visible
+                    .filter(Reference.to_user_id == request.to_user_id)
+                )
```

Comment:
Not sure why I can't comment here, but I think the subquery also needs a visibility check? (`sub = session...`)

Reply:
I don't think so. Because in the first query we filter out everything sent to/from banned users. Even if we pick it up in the second query, the outerjoin will get rid of it. We can add it, but it's an extra join. I think not adding is faster (and maybe clearer?)

Reply:
Ah, you're right.

You have to move these joins out of the if queries then, into line 58. You don't only want to filter out banned stuff if you're filtering for a user, they're separate things.

Reply:
I don't understand. The joins are dependent upon if the request wants to see references about a user or written by the user? Where would you move them?

(Also, I assume you mean lines 59 and 60?)